File: app/src/main/java/com/pomoremote/MainActivity.java
package com.pomoremote;

import android.content.BroadcastReceiver;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.ServiceConnection;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.widget.Button;
import android.widget.TextView;

import androidx.appcompat.app.AppCompatActivity;

import com.pomoremote.service.PomodoroService;
import com.pomoremote.timer.TimerState;

import java.util.Locale;

public class MainActivity extends AppCompatActivity {
    private PomodoroService service;
    private boolean isBound = false;
    
    private TextView tvTimer;
    private TextView tvPhase;
    private TextView tvStatus;
    private TextView tvProgress;
    private Button btnToggle;

    private final ServiceConnection connection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder binder) {
            PomodoroService.LocalBinder localBinder = (PomodoroService.LocalBinder) binder;
            service = localBinder.getService();
            isBound = true;
            updateUI(service.getCurrentState());
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {
            isBound = false;
        }
    };

    private final BroadcastReceiver stateReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (isBound && service != null) {
                updateUI(service.getCurrentState());
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvTimer = findViewById(R.id.tvTimer);
        tvPhase = findViewById(R.id.tvPhase);
        tvStatus = findViewById(R.id.tvConnectionStatus);
        tvProgress = findViewById(R.id.tvProgress);
        btnToggle = findViewById(R.id.btnToggle);
        
        findViewById(R.id.btnSettings).setOnClickListener(v -> 
            startActivity(new Intent(this, SettingsActivity.class)));
            
        btnToggle.setOnClickListener(v -> {
            if (isBound && service != null) service.toggleTimer();
        });
        
        findViewById(R.id.btnSkip).setOnClickListener(v -> {
            if (isBound && service != null) service.skipTimer();
        });
        
        findViewById(R.id.btnReset).setOnClickListener(v -> {
            if (isBound && service != null) service.resetTimer();
        });

        startService();
    }

    private void startService() {
        Intent intent = new Intent(this, PomodoroService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(intent);
        } else {
            startService(intent);
        }
        bindService(intent, connection, Context.BIND_AUTO_CREATE);
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(stateReceiver, new IntentFilter("com.pomoremote.STATE_UPDATE"), Context.RECEIVER_NOT_EXPORTED);
        } else {
            registerReceiver(stateReceiver, new IntentFilter("com.pomoremote.STATE_UPDATE"));
        }
        
        if (isBound && service != null) {
            service.connect();
            updateUI(service.getCurrentState());
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        unregisterReceiver(stateReceiver);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (isBound) {
            unbindService(connection);
            isBound = false;
        }
    }

    private void updateUI(TimerState state) {
        if (state == null) return;

        int minutes = (int) state.remaining / 60;
        int seconds = (int) state.remaining % 60;
        tvTimer.setText(String.format(Locale.US, "%02d:%02d", minutes, seconds));

        String phaseName = state.phase;
        if (TimerState.PHASE_WORK.equals(state.phase)) phaseName = "Focus";
        else if (TimerState.PHASE_SHORT.equals(state.phase)) phaseName = "Short Break";
        else if (TimerState.PHASE_LONG.equals(state.phase)) phaseName = "Long Break";
        tvPhase.setText(phaseName);
        
        tvProgress.setText("Completed: " + state.completed);

        if (TimerState.STATUS_RUNNING.equals(state.status)) {
            btnToggle.setText("Pause");
        } else {
            btnToggle.setText("Start");
        }
        
        if (service != null) {
            if (service.isConnected()) {
                tvStatus.setText("Connected");
                tvStatus.setTextColor(getResources().getColor(R.color.status_connected));
            } else {
                tvStatus.setText("Offline");
                tvStatus.setTextColor(getResources().getColor(R.color.status_offline));
            }
        }
    }
}

File: app/src/main/java/com/pomoremote/MainActivity.kt
package com.pomoremote

import android.Manifest
import android.content.*
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.IBinder
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.navigation.fragment.NavHostFragment
import androidx.navigation.ui.setupWithNavController
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.pomoremote.service.PomodoroService
import com.pomoremote.ui.TimerFragment
import com.pomoremote.util.UtilPreferenceManager

class MainActivity : AppCompatActivity() {
    var service: PomodoroService? = null
        private set
    var isBound = false
        private set

    lateinit var prefs: UtilPreferenceManager
        private set

    private val connection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, binder: IBinder) {
            val localBinder = binder as PomodoroService.LocalBinder
            service = localBinder.service
            isBound = true
            service?.connect()
            updateCurrentFragment()
        }

        override fun onServiceDisconnected(name: ComponentName) {
            isBound = false
            service = null
        }
    }

    private val stateReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            updateCurrentFragment()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        prefs = UtilPreferenceManager(this)

        // Setup Navigation
        val navHostFragment = supportFragmentManager
            .findFragmentById(R.id.nav_host_fragment) as NavHostFragment
        val navController = navHostFragment.navController
        val navView: BottomNavigationView = findViewById(R.id.nav_view)
        navView.setupWithNavController(navController)

        startService()
        requestNotificationPermission()
    }

    private fun updateCurrentFragment() {
        val navHostFragment = supportFragmentManager.findFragmentById(R.id.nav_host_fragment) as? NavHostFragment
        val currentFragment = navHostFragment?.childFragmentManager?.primaryNavigationFragment

        if (currentFragment is TimerFragment) {
            service?.currentState?.let { currentFragment.updateUI(it) }
        }
    }

    fun toggleTimer() {
        if (isBound) service?.toggleTimer()
    }

    fun skipTimer() {
        if (isBound) service?.skipTimer()
    }

    fun resetTimer() {
        if (isBound) service?.resetTimer()
    }

    private fun requestNotificationPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS)
                != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(
                    this,
                    arrayOf(Manifest.permission.POST_NOTIFICATIONS),
                    NOTIFICATION_PERMISSION_REQUEST_CODE
                )
            }
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        // Handle permissions if needed
    }

    private fun startService() {
        val intent = Intent(this, PomodoroService::class.java)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForegroundService(intent)
        } else {
            startService(intent)
        }
        bindService(intent, connection, Context.BIND_AUTO_CREATE)
    }

    override fun onResume() {
        super.onResume()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            registerReceiver(stateReceiver, IntentFilter("com.pomoremote.STATE_UPDATE"), Context.RECEIVER_NOT_EXPORTED)
        } else {
            registerReceiver(stateReceiver, IntentFilter("com.pomoremote.STATE_UPDATE"))
        }

        if (isBound) {
            service?.connect()
            updateCurrentFragment()
        }
    }

    override fun onPause() {
        super.onPause()
        unregisterReceiver(stateReceiver)
    }

    override fun onDestroy() {
        super.onDestroy()
        if (isBound) {
            unbindService(connection)
            isBound = false
        }
    }

    companion object {
        private const val NOTIFICATION_PERMISSION_REQUEST_CODE = 1
    }
}
